### PR TITLE
OpenMPTarget init-join fix

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR adds verification for OpenMP process and initialization in init-join reducers.
This also fixes the backward_compatibility test failure with the backend.

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
How is calling two functions that check preconditions fixing anything?

**Comment by dalg24** [comment]
If the issue is related to setting the value of `MAX_ACTIVE_THREADS` https://github.com/kokkos/kokkos/blob/6a95b5f3a7f9382d2251fec69dcbfe5cfefa8acc/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.cpp#L70-L71
why is initialization in `OpenMPTargetExec::verify_initialized` the proper place to do that?
Why can't it be done at initialization of the backend?

**Comment by rgayatri23** [comment]
> If the issue is related to setting the value of `MAX_ACTIVE_THREADS`
> 
> https://github.com/kokkos/kokkos/blob/6a95b5f3a7f9382d2251fec69dcbfe5cfefa8acc/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.cpp#L70-L71
> 
> 
> why is initialization in `OpenMPTargetExec::verify_initialized` the proper place to do that?
> Why can't it be done at initialization of the backend?

Yes the issue is related to `MAX_ACTIVE_THREADS`. That variable is used to calculate maximum in flight teams which in-turn determines the scratch space allocated. 
The variable can be set during the backend initialization, which is probably where it should go I guess if we only want to do it only once.

**Comment by masterleinad** [comment]
> The variable can be set during the backend initialization, which is probably where it should go I guess if we only want to do it only once.

Would there be a reason to do it more than once?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp [hunk 1]
```diff
@@ -197,6 +197,10 @@ struct ParallelReduceSpecialize<FunctorType, Kokkos::RangePolicy<PolicyArgs...>,
 
   static void execute_init_join(const FunctorType& f, const PolicyType& p,
                                 PointerType ptr, const bool ptr_on_device) {
+    OpenMPTargetExec::verify_is_process(
+        "Kokkos::Experimental::OpenMPTarget parallel_for");
+    OpenMPTargetExec::verify_initialized(
+        "Kokkos::Experimental::OpenMPTarget parallel_for");
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp:203]: Are these changes still necessary?

Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp]: Firstly, the message label was wrong which I fixed now.
Secondly, I think yes , we are checking for those in other similar routines too. Not having here was an oversight.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp [hunk 2]
```diff
@@ -84,9 +84,9 @@ struct ParallelReduceSpecialize<FunctorType, Kokkos::RangePolicy<PolicyArgs...>,
   static void execute_reducer(const FunctorType& f, const PolicyType& p,
                               PointerType result_ptr, bool ptr_on_device) {
     OpenMPTargetExec::verify_is_process(
-        "Kokkos::Experimental::OpenMPTarget parallel_for");
+        "Kokkos::Experimental::OpenMPTarget parallel_reduce");
     OpenMPTargetExec::verify_initialized(
-        "Kokkos::Experimental::OpenMPTarget parallel_for");
+        "Kokkos::Experimental::OpenMPTarget parallel_reduce");
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp:89]: Wouldn't it be better to call these functions in `execute` directly before the dispatch to `execute_init_join`, `execute_reducer`,...? In that case, you could also be more precise about the policy used in the error message.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.cpp [hunk 3]
```diff
@@ -115,6 +115,9 @@ void OpenMPTargetInternal::impl_finalize() {
 void OpenMPTargetInternal::impl_initialize() {
   m_is_initialized = true;
 
+  Kokkos::Impl::OpenMPTargetExec::MAX_ACTIVE_THREADS =
+      Kokkos::Experimental::OpenMPTarget().concurrency();
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.cpp:119]:
```suggestion
  Kokkos::Impl::OpenMPTargetExec::MAX_ACTIVE_THREADS =
      concurrency();
```

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp (hunk 2 above)
Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp]: Edited the error message.

**Review by masterleinad — APPROVED** [review]
OK with me.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
The openmptarget build passed.  Not waiting on the rest

**Comment by dalg24** [comment]
@rgayatri23 make sure you add to the changelog
When was this bug introduced? Did you investigate?

**Comment by rgayatri23** [comment]
> @rgayatri23 make sure you add to the changelog When was this bug introduced? Did you investigate?

Yeah this was introduced in when we did #6043
